File: sdk/javascript/src/media.ts
// Multimodal media wrappers over pie:core/media.

import {
  Image as _Image,
  Video as _Video,
  Audio as _Audio,
} from 'pie:core/media';

import type { Model } from './model.js';

function asBytes(bytes: Uint8Array | ArrayBuffer | Iterable<number>): Uint8Array {
  if (bytes instanceof Uint8Array) return bytes;
  if (bytes instanceof ArrayBuffer) return new Uint8Array(bytes);
  return new Uint8Array(Array.from(bytes));
}

/** Host-side preprocessed still image. */
export class Image {
  /** @internal */
  readonly _handle: _Image;

  /** @internal */
  constructor(handle: _Image) {
    this._handle = handle;
  }

  /** Decode image bytes for `model`. */
  static fromBytes(model: Model, bytes: Uint8Array | ArrayBuffer | Iterable<number>): Image {
    return new Image(_Image.fromBytes(model._handle, asBytes(bytes)));
  }

  /** Soft-token rows / KV slots occupied by this span. */
  tokenCount(): number { return this._handle.tokenCount(); }

  /** Sequence-position span consumed by this image. */
  positionSpan(): number { return this._handle.positionSpan(); }

  /** `(t, h, w)` merged-token grid. */
  grid(): [number, number, number] { return this._handle.grid(); }

  /** Model-specific delimiter tokens before the image span. */
  prefixTokens(): Uint32Array { return this._handle.prefixTokens(); }

  /** Model-specific delimiter tokens after the image span. */
  suffixTokens(): Uint32Array { return this._handle.suffixTokens(); }
}

/** Host-side decoded and sampled video clip. */
export class Video {
  /** @internal */
  readonly _handle: _Video;

  /** @internal */
  constructor(handle: _Video) {
    this._handle = handle;
  }

  /** Decode and uniformly sample up to `maxFrames` frames for `model`. */
  static fromBytes(
    model: Model,
    bytes: Uint8Array | ArrayBuffer | Iterable<number>,
    maxFrames: number,
  ): Video {
    return new Video(_Video.fromBytes(model._handle, asBytes(bytes), maxFrames));
  }

  /** Number of sampled frames. */
  frameCount(): number { return this._handle.frameCount(); }

  /** The `index`-th sampled frame as an image span. */
  frame(index: number): Image { return new Image(this._handle.frame(index)); }

  /** Timestamp in seconds for the `index`-th sampled frame. */
  timestamp(index: number): number { return this._handle.timestamp(index); }
}

/** Host-side preprocessed audio clip. */
export class Audio {
  /** @internal */
  readonly _handle: _Audio;

  /** @internal */
  constructor(handle: _Audio) {
    this._handle = handle;
  }

  /** Decode audio bytes for `model`. */
  static fromBytes(model: Model, bytes: Uint8Array | ArrayBuffer | Iterable<number>): Audio {
    return new Audio(_Audio.fromBytes(model._handle, asBytes(bytes)));
  }

  /** Soft-token rows / KV slots occupied by this clip. */
  tokenCount(): number { return this._handle.tokenCount(); }

  /** Sequence-position span consumed by this audio clip. */
  positionSpan(): number { return this._handle.positionSpan(); }

  /** Model-specific delimiter tokens before the audio span. */
  prefixTokens(): Uint32Array { return this._handle.prefixTokens(); }

  /** Model-specific delimiter tokens after the audio span. */
  suffixTokens(): Uint32Array { return this._handle.suffixTokens(); }
}

File: sdk/inferlet/python/src/inferlet/media.py
"""
Multimodal input — ``pie:inferlet/media``.

The inferlet hands the host raw encoded bytes (:meth:`Image.from_bytes` for
PNG/JPEG, :meth:`Video.from_bytes` for animated GIF, :meth:`Audio.from_bytes`
for WAV) and the host decodes + preprocesses per the bound model. A span
enters the sequence as the token run its handle answers, and the handle
crosses again beside the tokens (``ForwardPass.media([img])``) to carry the
payload::

    img = media.Image.from_bytes(png)
    toks = model.encode("Describe: ") + img.tokens() + model.encode(" briefly.")
    fwd.embed(tokens_ch, indptr_ch)
    fwd.media([img])
"""

from __future__ import annotations

from dataclasses import dataclass

from componentize_py_types import Err as _WitErr
from wit_world.imports import media as _media


class MediaError(Exception):
    """Bytes the host could not decode as the requested medium."""


@dataclass(frozen=True)
class MergedGrid:
    t: int
    h: int
    w: int


class _Span:
    __slots__ = ("_inner",)

    def __init__(self, inner) -> None:
        self._inner = inner

    @property
    def handle(self):
        """The WIT resource, for ``ForwardPass.media``."""
        return self._inner

    def tokens(self) -> list[int]:
        """The placeholder run this span enters the sequence as."""
        return list(self._inner.tokens())

    def digest(self) -> bytes:
        return bytes(self._inner.digest())

    def token_count(self) -> int:
        return self._inner.token_count()

    def position_span(self) -> int:
        return self._inner.position_span()

    def prefix_tokens(self) -> list[int]:
        return list(self._inner.prefix_tokens())

    def suffix_tokens(self) -> list[int]:
        return list(self._inner.suffix_tokens())


class Image(_Span):
    @classmethod
    def from_bytes(cls, data: bytes) -> "Image":
        try:
            return cls(_media.Image.from_bytes(bytes(data)))
        except _WitErr as e:
            raise MediaError(f"image: {e.value}") from None

    def grid(self) -> MergedGrid:
        g = self._inner.grid()
        return MergedGrid(g.t, g.h, g.w)


class Audio(_Span):
    @classmethod
    def from_bytes(cls, data: bytes) -> "Audio":
        try:
            return cls(_media.Audio.from_bytes(bytes(data)))
        except _WitErr as e:
            raise MediaError(f"audio: {e.value}") from None


class Video:
    __slots__ = ("_inner",)

    def __init__(self, inner) -> None:
        self._inner = inner

    @classmethod
    def from_bytes(cls, data: bytes, max_frames: int) -> "Video":
        try:
            return cls(_media.Video.from_bytes(bytes(data), max_frames))
        except _WitErr as e:
            raise MediaError(f"video: {e.value}") from None

    def frame_count(self) -> int:
        return self._inner.frame_count()

    def frame(self, index: int) -> Image:
        try:
            return Image(self._inner.frame(index))
        except _WitErr as e:
            raise MediaError(f"video frame {index}: {e.value}") from None

    def timestamp(self, index: int) -> float:
        return self._inner.timestamp(index)
